### PR TITLE
[TOSA] Add torch.prim.NumToTensor.Scalar float support

### PR DESCRIPTION
**Pull request by AmosLewis**
Find this in OPT model : https://github.com/nod-ai/SHARK/issues/865
> Also find this f64 support issue in GPT2 and distilGPT2 https://github.com/nod-ai/SHARK/issues/494 with transformer 4.25.1

This patch needs the f64 support of tosa https://reviews.llvm.org/D142599

For f64 support test file and debug output: https://gist.github.com/AmosLewis/fca6b0d16ee325fcf7ee400459f4fd40

Related issues: https://github.com/llvm/torch-mlir/issues/1615

### CODE REVIEW

**Comment by AmosLewis** [comment]
@eric-k256 @ramiro050 is there any possibility to add f64 in tosa. I got it from GPT2/distilGPT2 model torchscript with transformer package 4.25.2. I have no idea how to walk around this unless I downgrade transformer version to 4.21.

**Review by ramiro050 — COMMENTED** [review]
I would've expected the e2e test for `NumToTensor` to pass on TOSA now

**Comment by eric-k256** [comment]
For TOSA, we're looking into adding something like a 'compiler profile' that would allow a wider set of data types, with a corresponding validation pass that would fail networks that are using f64/i64 with devices that don't expect it. If we added this to TOSA, and expanded the types in the dialect, would this be enough for you? Also, are your passes that consume TOSA capable of handling f64? I'd still recommend against f64 unless it is requested by the network developer, as it's likely to be slower on almost every device.

**Comment by AmosLewis** [comment]
> For TOSA, we're looking into adding something like a 'compiler profile' that would allow a wider set of data types, with a corresponding validation pass that would fail networks that are using f64/i64 with devices that don't expect it. If we added this to TOSA, and expanded the types in the dialect, would this be enough for you? Also, are your passes that consume TOSA capable of handling f64? I'd still recommend against f64 unless it is requested by the network developer, as it's likely to be slower on almost every device.

What I need is at least if we use cast, it can walk around this issue.Fot i64 I use to use tosa::cast i64 to i32, then at the end of the work, I use tosa::cast i32 back to i64. It work. But for f64, cast fail. As you can see in the following:
With tosa::cast f32 to f64 code(has been deleted in this patch):
```
  auto outElemTy = resultType.getElementType();
  if (outElemTy.isF64()) {
    auto resultF32 = tosa::getConstTensor<float>(rewriter, op, floatValue, {}).value();
    rewriter.replaceOpWithNewOp<tosa::CastOp>(op, resultType, resultF32);
  }
```
Bug Output: 
```
// *** IR Dump After Pattern Application ***
mlir-asm-printer: Verifying operation: func.func
ImplicitTypeIDRegistry::lookupOrInsert(mlir::InferShapedTypeOpInterface::Trait<Empty>)
'tosa.cast' op result #0 must be tensor of number values, but got 'tensor<f64>'
mlir-asm-printer: 'func.func' failed to verify and will be printed in generic form
"func.func"() ({
  %0 = "torch.constant.float"() {value = 8.000000e+00 : f64} : () -> !torch.float
  %1 = "builtin.unrealized_conversion_cast"(%0) : (!torch.float) -> f64
  %2 = "tosa.const"() {value = dense<8.000000e+00> : tensor<f32>} : () -> tensor<f32>
  %3 = "tosa.cast"(%2) : (tensor<f32>) -> tensor<f64>
  %4 = "torch.prim.NumToTensor.Scalar"(%0) : (!torch.float) -> !torch.vtensor<[],f64>
  "func.return"(%4) : (!torch.vtensor<[],f64>) -> ()
}) {function_type = () -> !torch.vtensor<[],f64>, sym_name = "torch.prim.NumToTensor.Scalar"} : () -> ()


} -> SUCCESS
//===-------------------------------------------===//

//===-------------------------------------------===//
Legalizing operation : 'func.return'(0x55a53fab0c30) {
  "func.return"(%4) : (!torch.vtensor<[],f64>) -> ()

  * Fold {
  } -> FAILURE : unable to fold
} -> FAILURE : no matched legalization pattern
//===-------------------------------------------===//
** Insert  : 'torch_c.from_builtin_tensor'(0x55a53fb17c60)
ImplicitTypeIDRegistry::lookupOrInsert(mlir::detail::PreservedAnalyses::AllAnalysesType)
/tmp/NumToTensor.mlir:3:8: error:   'tosa.cast' op result #0 must be tensor of number values, but got 'tensor<f64>'
  %1 = "torch.prim.NumToTensor.Scalar"(%float8.000000e00) : (!torch.float) -> !torch.vtensor<[],f64>
       ^
/tmp/NumToTensor.mlir:3:8: note: see current operation: %2 = "tosa.cast"(%1) : (tensor<f32>) -> tensor<f64>

```

**Comment by AmosLewis** [comment]
> I would've expected the e2e test for `NumToTensor` to pass on TOSA now

I guess they pass is because by default it generated f32. So no f64 issue.

**Comment by AmosLewis** [comment]
Let's at least enable the f32 for now.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -3350,13 +3350,25 @@ LogicalResult ConvertAtenOp<PrimNumToTensorScalarOp>::matchAndRewrite(
...
+  auto isFloat = matchPattern(op.getA(), m_TorchConstantFloat(&doubleValue));
+  int64_t intValue;
+  auto isInt = matchPattern(op.getA(), m_TorchConstantInt(&intValue));
+  if (!isFloat && !isInt)
+    return rewriter.notifyMatchFailure(op,
+                                       "Unable to extract the scalar constant");
+  float floatValue = static_cast<float>(doubleValue);
+
+  auto outElemTy = resultType.getElementType();
+  if (outElemTy.isa<mlir::IntegerType>() || outElemTy.isF32()) {
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3363]: `PrimNumToTensorScalarOp` does not produce `f32` tensors.

```python
In [5]: torch.ops.prim.NumToTensor(0.0).dtype
Out[5]: torch.float64
```

so this is not the right lowering for this op. Also, e2e tests are needed anytime a new op is extended. See: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

> DO use an [end-to-end test](https://github.com/llvm/torch-mlir/blob/main/docs/adding_an_e2e_test.md) if you are implementing a new op or extending the support for an existing op.

**Review by ramiro050 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3363]: Oh, it's XFAIL - "NumToTensorFloatModule_basic"

**Comment by AmosLewis** [comment]
> For TOSA, we're looking into adding something like a 'compiler profile' that would allow a wider set of data types, with a corresponding validation pass that would fail networks that are using f64/i64 with devices that don't expect it. If we added this to TOSA, and expanded the types in the dialect, would this be enough for you? Also, are your passes that consume TOSA capable of handling f64? I'd still recommend against f64 unless it is requested by the network developer, as it's likely to be slower on almost every device.

@eric-k256  Is there any possibility to add at least f64 in` tosa.cast`? For the new model [facebook/opt-125m](https://huggingface.co/facebook/opt-125m) I am working on. I definitely have no way to walk around the f64 support. Something like: 
%2 = "tosa.cast"(%1) : (tensor<f32>) -> tensor<f64>
%3 = "tosa.cast"(%2) : (tensor<f64>) -> tensor<f32>

**Comment by eric-k256** [comment]
I would prefer to start with the smallest possible scope for f64. If we only want to support it for cast, we could create a new type that is used for the inputs and outputs of tosa.cast instead of Tosa_Tensor, it could be something like Tosa_CastableTensor, which included all types from Tosa_Tensor and f64. That would allow us to manage the use of f64, avoiding it for most cases.

**Comment by AmosLewis** [comment]
the f64 support of tosa https://reviews.llvm.org/D142599

**Comment by AmosLewis** [comment]
> I would prefer to start with the smallest possible scope for f64. If we only want to support it for cast, we could create a new type that is used for the inputs and outputs of tosa.cast instead of Tosa_Tensor, it could be something like Tosa_CastableTensor, which included all types from Tosa_Tensor and f64. That would allow us to manage the use of f64, avoiding it for most cases.

Done. Please review in llvm

**Comment by AmosLewis** [comment]
```
module attributes {torch.debug_module_name = "NumToTensorFloatModule"} {
  func.func @forward() -> !torch.vtensor<[],f64> {
    %float1.000000e00 = torch.constant.float 1.000000e+00 loc(#loc1)
    %0 = torch.prim.NumToTensor.Scalar %float1.000000e00 : !torch.float -> !torch.vtensor<[],f64> loc(#loc2)
    return %0 : !torch.vtensor<[],f64> loc(#loc)
  } loc(#loc)
} loc(#loc)
```


```
module attributes {torch.debug_module_name = "NumToTensorFloatModule"} {
  func.func @forward() -> !torch.vtensor<[],f64> {
    %float1.000000e00 = torch.constant.float 1.000000e+00
    %0 = "tosa.const"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %1 = "tosa.cast"(%0) : (tensor<f32>) -> tensor<f64>
    %2 = torch_c.from_builtin_tensor %1 : tensor<f64> -> !torch.vtensor<[],f64>
    return %2 : !torch.vtensor<[],f64>
  }
}
```

**Comment by AmosLewis** [comment]
Since this https://reviews.llvm.org/rGa2dcd994a7f8cc33640f58105276b78acf3483e5 has been merged and update in torch-mlir. This patch should be ok to merged now.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (ramiro050) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3598]: I don't see this as having been addressed

**Comment by sjw36** [comment]
I am trying to solve a similar problem with f64 vtensor.literals in the dynamo flow. The rewrite already has to convert signed integers to unsigned, pretty simple to also convert f64 to f32. 
And in general if TOSA is not going to support F64, would it be valid to make the typeConverter convert to f32?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3598]:
```
module attributes {torch.debug_module_name = "NumToTensorFloatModule"} {
  func.func @forward() -> !torch.vtensor<[],f64> {
    %float1.000000e00 = torch.constant.float 1.000000e+00 loc(#loc1)
    %0 = torch.prim.NumToTensor.Scalar %float1.000000e00 : !torch.float -> !torch.vtensor<[],f64> loc(#loc2)
    return %0 : !torch.vtensor<[],f64> loc(#loc)
  } loc(#loc)
} loc(#loc)
```
`➜  torch-mlir git:(indexput) ✗ torch-mlir-opt -convert-torch-to-tosa /tmp/NumToTensorFloatModule.mlir       `
```
 module attributes {torch.debug_module_name = "NumToTensorFloatModule"} {
  func.func @forward() -> !torch.vtensor<[],f64> {
    %float1.000000e00 = torch.constant.float 1.000000e+00
    %0 = "tosa.const"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %1 = "tosa.cast"(%0) : (tensor<f32>) -> tensor<f64>
    %2 = torch_c.from_builtin_tensor %1 : tensor<f64> -> !torch.vtensor<[],f64>
    return %2 : !torch.vtensor<[],f64>
  }
}
```

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3598]: Got it, Need to change the tosa::constOp to support Tosa_Tensor_Cast f64.
def Tosa_ConstOp : Tosa_Op<"const", ...
  let results = (outs
    Tosa_Tensor_Cast:$output
  );

**Comment by AmosLewis** [comment]
> I am trying to solve a similar problem with f64 vtensor.literals in the dynamo flow. The rewrite already has to convert signed integers to unsigned, pretty simple to also convert f64 to f32. And in general if TOSA is not going to support F64, would it be valid to make the typeConverter convert to f32?

Just one line change the tosa::constOp definition to support Tosa_Tensor_Cast f64 should be enough.
def Tosa_ConstOp : Tosa_Op<"const", ...
let results = (outs
**Tosa_Tensor_Cast**:$output
);

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3731]: Done

Comment (AmosLewis) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3731]: Done

**Comment by AmosLewis** [comment]
https://reviews.llvm.org/D145336 this has been merged and update in torch-mlir. Please review @ramiro050

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -3718,13 +3718,27 @@ LogicalResult ConvertAtenOp<PrimNumToTensorScalarOp>::matchAndRewrite(
...
-    return rewriter.notifyMatchFailure(
-        op, "unimplemented: input should be a torch constant int");
+  double doubleValue;
+  auto isFloat = matchPattern(op.getA(), m_TorchConstantFloat(&doubleValue));
+  int64_t intValue;
+  auto isInt = matchPattern(op.getA(), m_TorchConstantInt(&intValue));
+  if (!isFloat && !isInt)
+    return rewriter.notifyMatchFailure(op,
+                                       "Unable to extract the scalar constant");
+  float floatValue = static_cast<float>(doubleValue);
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3728]: This should not be casted to lower precision. This leads to a loss of information

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 3]
```diff
@@ -3718,13 +3718,27 @@ LogicalResult ConvertAtenOp<PrimNumToTensorScalarOp>::matchAndRewrite(
...
+
+  auto outElemTy = resultType.getElementType();
+  if (outElemTy.isa<mlir::IntegerType>() || outElemTy.isF32()) {
+    DenseElementsAttr constAttr =
+        isInt ? DenseElementsAttr::get(resultType, {intValue})
+              : DenseElementsAttr::get(resultType, {floatValue});
+    rewriter.replaceOpWithNewOp<tosa::ConstOp>(op, resultType, constAttr);
+  } else if (outElemTy.isF64()) {
+    auto resultF32 =
+        tosa::getConstTensor<float>(rewriter, op, floatValue, {}).value();
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3738]: this should be an f64 constant

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 4]
```diff
@@ -3718,13 +3718,27 @@ LogicalResult ConvertAtenOp<PrimNumToTensorScalarOp>::matchAndRewrite(
...
+  auto isFloat = matchPattern(op.getA(), m_TorchConstantFloat(&doubleValue));
+  int64_t intValue;
+  auto isInt = matchPattern(op.getA(), m_TorchConstantInt(&intValue));
+  if (!isFloat && !isInt)
+    return rewriter.notifyMatchFailure(op,
+                                       "Unable to extract the scalar constant");
+  float floatValue = static_cast<float>(doubleValue);
+
+  auto outElemTy = resultType.getElementType();
+  if (outElemTy.isa<mlir::IntegerType>() || outElemTy.isF32()) {
```
Comment (ramiro050) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3731]: outElemTy is never f32 for this op

**Review by ramiro050 — CHANGES_REQUESTED** [review]

**Review by ramiro050 — APPROVED** [review]

**Pull request merged**